apps/src/dance/LottieDancerRenderer.ts

```diff
@@ -0,0 +1,673 @@
+        const pathStr = [...pathNames, shapeNode.nm || '']
+          .filter(Boolean)
+          .join(' / ');
+
+        // 1) Accessories FIRST (so they override hose matches)
+        const matchesSecondary =
+          segs.some(s => anyMatch(s, SECONDARY_RXS)) ||
+          anyMatch(shapeName, SECONDARY_RXS) ||
+          anyMatch(layerName, SECONDARY_RXS) ||
+          anyMatch(pathStr, SECONDARY_RXS);
+
+        const matchesTertiary =
+          segs.some(s => anyMatch(s, TERTIARY_RXS)) ||
+          anyMatch(shapeName, TERTIARY_RXS) ||
+          anyMatch(layerName, TERTIARY_RXS) ||
+          anyMatch(pathStr, TERTIARY_RXS);
+
+        if (matchesSecondary) {
+          paintNode(shapeNode, palette.secondary);
+          continue;
+        }
+        if (matchesTertiary) {
+          paintNode(shapeNode, palette.tertiary);
+          continue;
+        }
+
+        // 2) Rubber-hose lines/areas on arms/legs → PRIMARY (if not accessory)
+        const inArmLegContext =
+          layerSegs.some(isArmLegSeg) || segs.some(isArmLegSeg);
+        const inHoseGroup = segs.some(isHoseySeg);
+        const ty = shapeNode.ty;
+        if ((ty === 'st' || ty === 'fl') && inArmLegContext && inHoseGroup) {
+          paintNode(shapeNode, palette.primary);
+          continue;
+        }
+      }
+    };
+
+    const visitLayer = (layer?: LottieLayer) => {
+      if (!layer) return;
+      if (Array.isArray(layer.shapes)) {
+        walkShapes(layer.shapes, layer.nm || '', []);
+      }
+    };
+
+    animationData.layers?.forEach(visitLayer);
+    animationData.assets?.forEach(a =>
+      (a as LottieAssetPrecomp).layers?.forEach(visitLayer)
+    );
+  }
+
+  /** Loads head.png as a data URL and detects natural size (falls back to 1000×1000). */
+  private async _fetchHeadImageInfo(): Promise<HeadImageInfo | null> {
+    const headUrl = this._resolveHeadUrl();
+    try {
+      const res = await fetch(headUrl, {cache: 'no-cache'});
+      if (!res.ok) return null;
+
+      const blob = await res.blob();
+
+      const dataUrl = await new Promise<string>(resolve => {
+        const fr = new FileReader();
+        fr.onload = () => resolve(fr.result as string);
+        fr.readAsDataURL(blob);
+      });
+
+      const {width, height} = await new Promise<{
+        width: number;
+        height: number;
+      }>(resolve => {
+        const img = new Image();
+        img.onload = () =>
+          resolve({
+            width: img.naturalWidth || 1000,
+            height: img.naturalHeight || 1000,
+          });
+        img.onerror = () => resolve({width: 1000, height: 1000});
+        img.src = dataUrl;
+      });
+
+      return {dataUrl, w: width, h: height};
+    } catch {
+      return null;
+    }
+  }
+
+  /**
+   * Searches all precomp assets for a layer named like “head”.
+   * Returns pointers so we can locate the child comp that actually draws the head.
+   */
+  private _findHeadPrecompLayerDeep(animationData: LottieJSON): {
```

Comment:
I wonder if we might want to move these deeper Lottie-specific utility functions into their own helper file?  This will help separate out the complex Lottie-specific work from the more general dancer renderer.